File: dyros_robot_controller/drc/differential_wheel.py
from .dyros_robot_controller_wrapper_cpp import DifferentialWheel as DifferentialWheelcpp
import numpy as np

class DifferentialWheel:
    """
    DifferentialWheel class for controlling a differential-drive robot.

    Attributes:
        wheel_names (list of str): Names of the wheels.
        base_width (float): Width of the robot's base.
        wheel_radius (float): Radius of the wheels.
        lin_vel_limit (float): Linear velocity limit.
        ang_vel_limit (float): Angular velocity limit.
        lin_acc_limit (float): Linear acceleration limit.
        ang_acc_limit (float): Angular acceleration limit.
        hz (float): Control frequency.

    Methods:
        IK(desired_base_velocity: np.array) -> np.array:
            Computes the inverse kinematics for the desired base velocity.
        
        FK(desired_wheel_vel: np.array) -> np.array:
            Computes the forward kinematics for the given wheel velocities.
        
        VelocityCommand(desired_base_vel: np.array, current_base_vel: np.array) -> np.array:
            Computes the velocity command for the wheels based on desired and current base velocities.
    """
    
    def __init__(self, wheel_names, base_width, wheel_radius, lin_vel_limit, ang_vel_limit, lin_acc_limit, ang_acc_limit, control_freq):
        self.wheel_names = wheel_names
        self.base_width = base_width
        self.wheel_radius = wheel_radius
        self.lin_vel_limit = lin_vel_limit
        self.ang_vel_limit = ang_vel_limit
        self.lin_acc_limit = lin_acc_limit
        self.ang_acc_limit = ang_acc_limit
        self.hz = control_freq
        
        self.controller = DifferentialWheelcpp(self.wheel_names,
                                               self.base_width,
                                               self.wheel_radius,
                                               self.lin_vel_limit,
                                               self.ang_vel_limit,
                                               self.lin_acc_limit,
                                               self.ang_acc_limit,
                                               self.hz)
        
    def IK(self, desired_base_velocity: np.array) -> np.array:
        """
        Computes the inverse kinematics for the desired base velocity.
        
        Args:
            desired_base_velocity (np.array): Desired base velocity as a 3D vector.
            
        Returns:
            np.array: Wheel velocities.
        """
        assert isinstance(desired_base_velocity, np.ndarray), "Input must be a numpy array."
        try:
            return self.controller.IK(desired_base_velocity)
        except Exception as e:
            print(f"Error in IK calculation: {e}")
            return np.array([])

    def FK(self, desired_wheel_vel: np.array) -> np.array:
        """
        Computes the forward kinematics for the given wheel velocities.
        
        Args:
            desired_wheel_vel (np.array): Desired wheel velocities.
            
        Returns:
            np.array: Base velocities.
        """
        assert isinstance(desired_wheel_vel, np.ndarray), "Input must be a numpy array."
        try:
            return self.controller.FK(desired_wheel_vel)
        except Exception as e:
            print(f"Error in FK calculation: {e}")
            return np.array([])

    def VelocityCommand(self, desired_base_vel: np.array, current_base_vel: np.array) -> np.array:
        """
        Computes the velocity command for the wheels based on desired and current base velocities.
        
        Args:
            desired_base_vel (np.array): Desired base velocity.
            current_base_vel (np.array): Current base velocity.
            
        Returns:
            np.array: Wheel velocity commands.
        """
        assert isinstance(desired_base_vel, np.ndarray) and isinstance(current_base_vel, np.ndarray), \
            "Both inputs must be numpy arrays."
        try:
            desired_base_vel[1] = 0
            return self.controller.VelocityCommand(desired_base_vel, current_base_vel)
        except Exception as e:
            print(f"Error in VelocityCommand calculation: {e}")
            return np.array([])

File: dyros_robot_controller/src/DifferentialWheel.cpp
#include "dyros_robot_controller/DifferentialWheel.h"
#include <cassert>
#include <cmath>

namespace DyrosRobotController
{
    DifferentialWheel::DifferentialWheel(std::vector<std::string> wheel_names, double base_width, double wheel_radius, double lin_vel_limit, double ang_vel_limit, double lin_acc_limit, double ang_acc_limit, double control_freq)
        : wheel_names_(wheel_names), 
          base_width_(base_width), 
          wheel_radius_(wheel_radius), 
          lin_vel_limit_(lin_vel_limit), 
          ang_vel_limit_(ang_vel_limit),
          lin_acc_limit_(lin_acc_limit), 
          ang_acc_limit_(ang_acc_limit), 
          hz_(control_freq) 
    {

    }

    Eigen::VectorXd DifferentialWheel::IK(const Eigen::Vector3d& desired_base_velocity) 
    {
        assert(std::abs(desired_base_velocity[1]) < 1e-3 && "Desired velocity along y-axis must be zero!");

        Eigen::Vector2d base_velocity(desired_base_velocity[0], desired_base_velocity[2]);
        Eigen::Matrix2d transformation_matrix;
        transformation_matrix << 1,  base_width_ / 2,
                                 1, -base_width_ / 2;
        transformation_matrix /= wheel_radius_;

        Eigen::Vector2d desired_wheel_vel = transformation_matrix * base_velocity;
        Eigen::VectorXd desired_wheel_vel_align(wheel_names_.size());

        for (size_t i = 0; i < wheel_names_.size(); ++i) 
        {
            if (wheel_names_[i].find("right") != std::string::npos) 
            {
                desired_wheel_vel_align[i] = desired_wheel_vel[0];
            } 
            else 
            {
                desired_wheel_vel_align[i] = desired_wheel_vel[1];
            }
        }
        return desired_wheel_vel_align;
    }

    Eigen::Vector3d DifferentialWheel::FK(const Eigen::VectorXd& desired_wheel_vel)
    {
        assert(desired_wheel_vel.size() == wheel_names_.size());

        Eigen::Vector2d desired_wheel_vel_dealign;
        for (size_t i = 0; i < wheel_names_.size(); ++i) 
        {
            if (wheel_names_[i].find("right") != std::string::npos) 
            {
                desired_wheel_vel_dealign[0] = desired_wheel_vel[i];
            } 
            else
            {
                desired_wheel_vel_dealign[1] = desired_wheel_vel[i];
            }
        }

        Eigen::Matrix2d transformation_matrix;
        transformation_matrix << 0.5,              0.5,
                                 1 / base_width_, -1 / base_width_;
        transformation_matrix *= wheel_radius_;

        Eigen::Vector2d base_velocity = transformation_matrix * desired_wheel_vel_dealign;
        return Eigen::Vector3d(base_velocity[0], 0, base_velocity[1]);
    }

    Eigen::VectorXd DifferentialWheel::VelocityCommand(const Eigen::Vector3d& desired_base_vel, const Eigen::Vector3d& current_base_vel) 
    {
        Eigen::Vector3d desired_base_acc = (desired_base_vel - current_base_vel) * hz_;

        if (desired_base_acc.head<2>().norm() > lin_acc_limit_) 
        {
            desired_base_acc.head<2>() = lin_acc_limit_ * desired_base_acc.head<2>().normalized();
        }
        if (std::abs(desired_base_acc[2]) > ang_acc_limit_) 
        {
            desired_base_acc[2] = std::copysign(ang_acc_limit_, desired_base_acc[2]);
        }
        Eigen::Vector3d new_base_vel = current_base_vel + desired_base_acc / hz_;

        if (new_base_vel.head<2>().norm() > lin_vel_limit_) 
        {
            new_base_vel.head<2>() = lin_vel_limit_ * new_base_vel.head<2>().normalized();
        }
        if (std::abs(new_base_vel[2]) > ang_vel_limit_) 
        {
            new_base_vel[2] = std::copysign(ang_vel_limit_, new_base_vel[2]);
        }

        return IK(new_base_vel);
    }
} 

